archunit/src/main/java/com/tngtech/archunit/core/importer/AccessRecord.java

```diff
@@ -280,215 +274,131 @@ private static JavaCodeUnit getCaller(CodeUnit caller, ImportedClasses classes)
-                        return true;
-                    }
+        private static Optional<JavaField> searchFieldInInterfaces(JavaClass targetOwner, TargetInfo targetInfo) {
+            for (JavaClass rawInterface : targetOwner.getRawInterfaces()) {
+                Optional<JavaField> foundOnInterface = searchTargetField(rawInterface, targetInfo);
+                if (foundOnInterface.isPresent()) {
+                    return foundOnInterface;
                 }
-                return false;
             }
-        };
-
-        private static <T> Optional<T> uniqueTargetIn(Collection<T> collection) {
-            return collection.size() == 1 ? Optional.of(getOnlyElement(collection)) : Optional.<T>empty();
+            return Optional.empty();
         }
 
-        private static List<JavaClass> getArgumentTypesFrom(String descriptor, ImportedClasses classes) {
-            ImmutableList.Builder<JavaClass> result = ImmutableList.builder();
-            for (JavaClassDescriptor type : JavaClassDescriptorImporter.importAsmMethodArgumentTypes(descriptor)) {
-                result.add(classes.getOrResolve(type.getFullyQualifiedClassName()));
-            }
-            return result.build();
+        private static Optional<JavaField> searchFieldInSuperClass(JavaClass targetOwner, TargetInfo targetInfo) {
+            return targetOwner.getRawSuperclass().isPresent()
+                    ? searchTargetField(targetOwner.getRawSuperclass().get(), targetInfo)
+                    : Optional.<JavaField>empty();
         }
 
-        private static class ClassHierarchyPath implements Iterable<JavaClass> {
-            private final List<JavaClass> path;
+        private static Optional<JavaMethod> searchTargetMethod(JavaClass targetOwner, TargetInfo targetInfo) {
+            MatchingMethods matchingMethods = new MatchingMethods(targetInfo);
+            matchingMethods.addMatching(targetOwner.getMethods(), true);
+            return matchingMethods.hasMatch()
+                    // shortcut -> if we found it directly in the class we don't need to look further up the hierarchy
+                    ? matchingMethods.determineMostSpecificMethod()
+                    : searchTargetMethodInHierarchy(targetOwner, matchingMethods);
+        }
 
-            public ClassHierarchyPath(JavaClassDescriptor childType, JavaClass parent) {
-                Optional<JavaClass> child = tryFindChildInHierarchy(childType, parent);
-                path = child.isPresent() ? createPath(parent, child.get()) : Collections.<JavaClass>emptyList();
+        private static Optional<JavaMethod> searchTargetMethodInHierarchy(JavaClass targetOwner, MatchingMethods matchingMethods) {
+            Optional<JavaClass> superclass = targetOwner.getRawSuperclass();
+            if (superclass.isPresent()) {
+                matchingMethods.addMatching(superclass.get().getMethods(), true);
```

Comment:
I get why we should not include static methods but why only for interfaces?

Reply:
My guess is because for classes it's forbidden. I can see that the Reflection API only considers this case, but I can also see that it's impossible to define a static and a clashing instance method within the hierarchy of a class (feel free to prove me wrong, then I need to reconsider this :joy:)
I think the reason is that static methods on interfaces always needs to be called on the Interface. So you can't do

```
interface IFace {
  static void method() {}
}
class Clazz implements IFace {
  void call() {
    method(); // doesn't compile, only IFace.method() would...
  }
}
```

But for classes it's allowed to call them that way

```
class Parent {
  static void method() {}
}
class Child extends Parent {
  void call() {
    method(); // this is fine now
  }
  void method() {} // but this causes a compile error because of ambiguity
}
```

Don't ask me why decisions were made that way, but that's how it is :man_shrugging: My guess would be because interface default methods came later, so the behavior for interfaces was the only way to add this in a backwards compatible way :man_shrugging:

Reply:
Thanks for the example. That helped me to understand it.